### PR TITLE
Add initial WebSocket based chat service

### PR DESCRIPTION
**Pull request by ideanl**
**Patch description**
This PR is the initial implementation of a WebSocket based chat service. Much of this code is identical or very similar to the messenger chat service, which will be generalized in #2079. This service is simplified from the messenger service, and does not yet include display names, attachments (stickers, images, etc.), or quick replies.

**Testing steps**
Ran the service using `run.py` similar to the Messenger service and provided first the `chatbot` example task and then the `overworld_demo`. Then connected via the internal WebSocket based web-app and began chatting. Tested pairing in `overworld_demo` as well as singleplayer in `chatbot`.

Example server run command (for chatbot task): `./parlai/chat_service/services/websocket/run.py --config-path parlai/chat_service/tasks/chatbot/config.yml`
To customize the listening port, specify the `--port` flag (defaults to 35496).

Connect via WebSocket at `ws://<YOUR DOMAIN>:<PORT>/websocket`. Example websocket frontend JS:
```
var ws = new WebSocket("ws://localhost:35496/websocket");
ws.onmessage = function(msg) {
    console.log("Received Message: " + msg.data)
}
ws.send("Hey");
```
Sample output from above: 
```
Received Message: Welcome to the overworld for the ParlAI messenger chatbot demo. Please type "begin" to start.
```

Components that could be generalized in #2079:

**In agents.py:**
- `set_stored_data`
- `get_new_act_message`
- `act_blocking`
- `episode_done`
- `mark_inactive`
- `act`, `observe`, and `put_data` could be generalized with small changes such as ensuring messenger specific fields such as quick_replies are not required.

**In websocket_manager.py**:
- `MessengerWorldRunner` import
- `AgentState` import
- `shared_utils` import
- most of `parse_config` is identical, the rest of it can be overridden (such as page_id for messenger)
- `check_timeout_in_pool`
- most of `remove_agent_from_pool` with some post-removal done by messenger
- `add_agent_to_pool`
- `_expire_all_conversations`
- `_get_unique_pool`
- `get_agent_state`
- `_get_agent`
- part of `_complete_setup`
- `_load_model`
-  In, `start_task` the loop body and the `_done_callback` are mostly identical. However, the `while self.running` part is not identical because the WebSocket implementation uses a periodic tornado callback rather than a while loop. The following line is only relevant to messenger:
      `agents = [a for a in agents if a.disp_id is not None]`
- `_on_new_message`, except the message format would need to be consistent. Messenger also has `typing_on` and `handle_bot_read` calls which are irrelevant to WebSockets.
- `_log_debug`

### CODE REVIEW

File: parlai/chat_service/services/websocket/agents.py [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+        """Send an agent a message through the websocket manager"""
+        logging.info(f"Sending new message: {act}")
+        if 'payload' in act:
+            raise ValueError("Payload not supported yet by websockets")
+        else:
+            msg = act['text']
+            self.manager.observe_message(self.id, msg)
+
+    def put_data(self, message):
+        """Put data into the message queue if it hasn't already been seen"""
```
Comment (klshuster) [inline parlai/chat_service/services/websocket/agents.py]: should we be checking that the message has not already been seen?

File: parlai/chat_service/services/websocket/agents.py [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+        """
+        # if this is the first act since last sent message start timing
+        if self.message_request_time is None:
+            self.message_request_time = time.time()
+
+        # If checking timeouts
+        if timeout:
+            # If time is exceeded, timeout
+            if time.time() - self.message_request_time > timeout:
+                return self.mark_inactive()
```
Comment (klshuster) [inline parlai/chat_service/services/websocket/agents.py]: should this be defined somewhere?

File: parlai/chat_service/services/websocket/sockets.py [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+    return str(uuid.uuid4())
+
+
+class MessageSocketHandler(WebSocketHandler):
+    subs = {}
+
+    def __init__(self, *args, **kwargs):
+        self.message_callback = kwargs.pop('message_callback', None)
+        self.sid = get_rand_id()
+        super(MessageSocketHandler, self).__init__(*args, **kwargs)
```
Comment (klshuster) [inline parlai/chat_service/services/websocket/sockets.py]: nit: can just call `super().__init__...`

File: parlai/chat_service/services/websocket/sockets.py [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+            logging.info(f"Current subscribers: {self.subs}")
+
+    def on_close(self):
+        """Runs when a socket is closed"""
+        del self.subs[self.sid]
+
+    def on_message(self, message):
+        """Callback that runs when a new message is received from a client"""
+        logging.info('websocket message from client: {}'.format(message))
+        self.message_callback(message, self.sid)
```
Comment (klshuster) [inline parlai/chat_service/services/websocket/sockets.py]: will this fail if there is no `self.message_callback`? should we provide a default one?

File: parlai/chat_service/services/websocket/websocket_manager.py [hunk 5]
```diff
@@ -0,0 +1,458 @@
...
+
+# TODO: Use a generalized World Runner module (Issue #2079)
+from parlai.chat_service.services.messenger.world_runner import MessengerWorldRunner
+import parlai.chat_service.services.messenger.shared_utils as shared_utils
+from parlai.chat_service.services.websocket.sockets import MessageSocketHandler
+from agents import WebsocketAgent
+import tornado
+from tornado.options import options
+
+PORT = 35496
```
Comment (klshuster) [inline parlai/chat_service/services/websocket/websocket_manager.py]: just a question - is this an arbitrary number?

File: parlai/chat_service/services/websocket/agents.py (hunk 1 above)
Comment (ideanl) [reply parlai/chat_service/services/websocket/agents.py]: Updated the docs to reflect that we don't do the check. 

`get_new_act_message` removes the message from the queue and the `on_new_message` callback for WebSockets is only be called once per message, so duplicate messages shouldn't be an issue.

File: parlai/chat_service/services/websocket/agents.py (hunk 2 above)
Comment (ideanl) [reply parlai/chat_service/services/websocket/agents.py]: Defined it as a `pass` as in the messenger implementation. Included this in the PR description as a function that can be generalized out.

File: parlai/chat_service/services/websocket/sockets.py (hunk 4 above)
Comment (ideanl) [reply parlai/chat_service/services/websocket/sockets.py]: Defined a default that outputs a warning

File: parlai/chat_service/services/websocket/websocket_manager.py (hunk 5 above)
Comment (ideanl) [reply parlai/chat_service/services/websocket/websocket_manager.py]: Yeah arbitrary default port number that can be specified via `--port`. I found this port number in other uses of WebSockets internally

**Comment by ideanl** [comment]
Updated the description with more information about running/using the service. Also included more details regarding what can be generalized w.r.t the messenger implementation (for #2079).

File: parlai/chat_service/services/websocket/websocket_manager.py (hunk 5 above)
Comment (ideanl) [reply parlai/chat_service/services/websocket/websocket_manager.py]: Removed the constant from the manager and instead defined it as a default in the argument parser

**Pull request merged**